### ISSUE TITLE
[question] How to put Greenmail's self-signed certificate into the Java Keystore

### ISSUE THREAD

**Issue by tobiasrodehutskors** [greenmail-mail-test/greenmail#163]
Hi!

I am currently trying to integrate Greenmail into an OSGi based environment. I primarily focus on using Greenmail as a dummy SMTPS server. Since Greenmail uses self-signed certificates to establish an SSL/TLS connection, you are normaly required to set the ssl.SocketFactory.provider to the DummySSLSocketFactory before trying to connect to the Greenmail server:

``` java
Security.setProperty("ssl.SocketFactory.provider", DummySSLSocketFactory.class.getName());
```

Unfortunatly, this is not feasable in an OSGi environment, as this will result in a call to Class.forName(...) to actually load the DummySSLSocketFactory. This will result in a ClassNotFoundException since OSGi and Java reflection does not work well together.

Therefore, I would like to avoid setting the DummySSLSocketFactory in this way and am looking for an alternative. I was thinking if it is possible to get the self-signed certificate from Greenmail and add it programatically to the Java keystore. This should enable my test to successfully establish a secured connection. Do you have any hints for me if this is possible? Can I retrieve the self-signed certificate?

Thanks in advance

Tobias

**Comment by marcelmay** [comment]
Tobias,

GreenMail directly uses DummySSLServerSocketFactory to create a socket.
Check [AbstractServer.openServerSocket](https://github.com/greenmail-mail-test/greenmail/blob/master/greenmail-core/src/main/java/com/icegreen/greenmail/server/AbstractServer.java#L63).

What happens if you start GreenMail configured for SMTPS as an OSGI service, can you provide the stack trace?

**Comment by buildscientist** [comment]
@marcelmay I think what @tobiasrodehutskors is looking for is related to #130 

I still need to push my code up - but I haven't finished wiring this up to our Configuration. 

So once this is done - in theory you wouldn't have to grab the JKS from GreenMail - you can use your own.

**Comment by tobiasrodehutskors** [comment]
@marcelmay The problem is not in the server part of greenmail, but in my test cases for my client code. I cannot set the ssl.SocketFactory.provider property, since the DummySSLSocketFactory will then be loaded via reflections which does not work in the OSGi setup. (If you are interested to know more, try googling OSGi + class.forname )

@buildscientist You are right, this looks indead promising. But still, this would require to setup an additional keystore in my case, wouln't it?

I  also had a look at the DummySSLServerSocketFactory and saw that it contains a method to retrieve the used keystore. However, this keystore only contains a key, but not a certificate and I don't know how this key can be added to Java's truststore to successfully establish a secured connection.

Just for reference, I currently use the following code in my mail client implementation, which will be enabled in test cases:

```Java
            if (trustAllCertificates) {
                server.property("mail.smtp.ssl.trust", "*");
            }
```

**Comment by marcelmay** [comment]
Tobias,

if you want to import the keystore ... just export it by writing the bytes of DummySSLServerSocketFactory.HARD_CODED_KEY_STORE into a file?

Or package a greenmail.jks with your key along the GreenMail server?

**Comment by buildscientist** [comment]
Closing as this will be resolved by #130
